graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,372 @@
+    // then its children will be treated like root paths, since there
+    // is no parent maxAge to inherit.
+
+    if (!fieldPolicy.hasMaxAge()
+        && ((unwrappedReturnType instanceof GraphQLCompositeType && !inheritMaxAge)
+            || parameters.getExecutionStepInfo().getPath().isRootPath())) {
+      fieldPolicy.restrict(defaultMaxAge);
+    }
+
+    return new SimpleInstrumentationContext<ExecutionResult>() {
+      @Override
+      public void onCompleted(ExecutionResult result, Throwable t) {
+        state.restrictCachePolicy(fieldPolicy);
+        super.onCompleted(result, t);
+      }
+    };
+  }
+
+  @Override
+  public CompletableFuture<ExecutionResult> instrumentExecutionResult(
+      ExecutionResult executionResult, InstrumentationExecutionParameters parameters) {
+    final CacheControlState state = parameters.getInstrumentationState();
+
+    if (!state.getOverallPolicyString().isPresent()) {
+      return super.instrumentExecutionResult(executionResult, parameters);
+    }
+
+    if (!executionResult.getErrors().isEmpty()) {
+      return super.instrumentExecutionResult(executionResult, parameters);
+    }
+
+    ExecutionResultImpl resultWithCacheControl =
+        ExecutionResultImpl.newExecutionResult()
+            .from(executionResult)
+            .addExtension(EXTENSION_KEY, state.getOverallPolicyString().get())
+            .build();
+
+    return CompletableFuture.completedFuture(resultWithCacheControl);
+  }
+
+  enum CacheControlScope {
+    PUBLIC,
+    PRIVATE
+  }
+
+  private static class CacheControlState implements InstrumentationState {
+    private final CacheControlPolicy overallPolicy = new CacheControlPolicy();
+
+    public Optional<String> getOverallPolicyString() {
+      return overallPolicy.getPolicy();
+    }
+
+    public void restrictCachePolicy(CacheControlPolicy policy) {
+      this.overallPolicy.restrict(policy);
+    }
+  }
+
+  private static class CacheControlPolicy implements InstrumentationState {
+    private Optional<Integer> maxAge = Optional.empty();
+    private Optional<CacheControlScope> scope = Optional.of(CacheControlScope.PUBLIC);
+
+    void restrict(CacheControlPolicy policy) {
+      if (policy.maxAge.isPresent()
+          && (!maxAge.isPresent() || policy.maxAge.get() < maxAge.get())) {
+        this.maxAge = policy.maxAge;
+      }
+
+      if (policy.scope.isPresent()
+          && (!scope.isPresent() || !scope.get().equals(CacheControlScope.PRIVATE))) {
+        this.scope = policy.scope;
+      }
+    }
+
+    void restrict(Hint hint) {
+      if (hint.maxAge.isPresent() && (!maxAge.isPresent() || hint.maxAge.get() < maxAge.get())) {
+        this.maxAge = hint.maxAge;
+      }
+
+      if (hint.scope.isPresent()
+          && (!scope.isPresent() || !scope.get().equals(CacheControlScope.PRIVATE))) {
+        this.scope = hint.scope;
+      }
+    }
+
+    void restrict(Integer maxAge) {
+      this.maxAge = Optional.of(maxAge);
```

Comment:
This method is doing a replace rather than a restrict. I see it is currently only called in a place where you've verified there's no max age already but it could be used elsewhere later... You could also just remove  it and have its caller call replace instead (or well, rename it to replace).

That said it's a little weird that the way this overloaded API works is reliant on the fact that all of the fields have different types.  Do we really need these single-field overloads? Vs setters with different names?

Reply:
fixed this to be a true restrict — i'm open to changing the API but it makes sense to me the way it currently is 🤷🏻